build_providers: pass through SNAPCRAFT_{BUILD,IMAGE}_INFO to container or VM

jhenstridge opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----
In order to include a `snap/manifest.yaml` in a snap, you can set the `SNAPCRAFT_BUILD_INFO` environment variable.  However, this doesn't work with build providers that perform a build within a container/VM, since the variable is not passed to the child Snapcraft instance.

This should fix [bug #1871770][1].

[1]: https://bugs.launchpad.net/snapcraft/+bug/1871770

sergiusens commented on tests/unit/build_providers/test_base_provider.py:
```diff
@@ -293,6 +293,7 @@ def test_clean_multiple_parts(self):
 
+    @patch.dict("os.environ", {}, clear=True)
     def test_passthrough_environment_flags_empty(self):
         provider = ProviderImpl(project=self.project, echoer=self.echoer_mock)
```
```suggestion
        self.useFixture(fixture_setup.CleanEnvironment())
        provider = ProviderImpl(project=self.project, echoer=self.echoer_mock)
```

sergiusens commented on tests/unit/build_providers/test_base_provider.py:
```diff
@@ -293,6 +293,7 @@ def test_clean_multiple_parts(self):
             ["snapcraft", "clean", "part1", "part2"]
         )
 
+    @patch.dict("os.environ", {}, clear=True)
```
mind using our already used fixture instead?

sergiusens commented on tests/unit/build_providers/test_base_provider.py:
```diff
@@ -334,6 +336,60 @@ def test_passthrough_environment_flags_non_string(self):
 
+    @patch.dict("os.environ", {"SNAPCRAFT_BUILD_INFO": "1"}, clear=True)
+    def test_passthrough_environment_flags_build_info(self):
+        provider = ProviderImpl(project=self.project, echoer=self.echoer_mock)
```
```suggestion
        self.useFixture(fixture_setup.CleanEnvironment())
        self.useFixture(fixtures.EnvironmentVariable("SNAPCRAFT_BUILD_INFO, "1"))
        provider = ProviderImpl(project=self.project, echoer=self.echoer_mock)
```

sergiusens commented on tests/unit/build_providers/test_base_provider.py:
```diff
@@ -334,6 +336,60 @@ def test_passthrough_environment_flags_non_string(self):
             ),
         )
 
+    @patch.dict("os.environ", {"SNAPCRAFT_BUILD_INFO": "1"}, clear=True)
```
mind using what I proposed instead?

sergiusens commented on snapcraft/internal/build_providers/_base_provider.py:
```diff
@@ -385,6 +385,13 @@ def _get_env_command(self) -> Sequence[str]:
             env_list.append(f"{key}={value}")
 
+        # Pass through build information environment variables.
+        for key in ["SNAPCRAFT_BUILD_INFO", "SNAPCRAFT_IMAGE_INFO"]:
```
I think @cjp256 would agree, how about we make these self.build_provider_flags instead? That's in `snapcraft.cli._options._PROVIDER_OPTIONS`

cjp256 replied:
Yes, something like adding the following to provider options should be sufficient.
```
    dict(
        param_decls="--generate-manifest",
        is_flag=True,
        help="Generate snap manifest.",
        envvar="SNAPCRAFT_BUILD)INFO",
        supported_providers=["host", "lxd", "managed-host", "multipass"],
        hidden=True,
    ),
    dict(
        param_decls="--manifest-image-information",
        metavar="<json string>",
        help="Set snap manifest image-info",
        envvar="SNAPCRAFT_IMAGE_INFO",
        supported_providers=["host", "lxd", "managed-host", "multipass"],
        hidden=True,
    ),
```

Tests won't be required as it's covered with the generic provider option handling.

jhenstridge replied:
I've updated the PR to make these options.

For compatibility with the existing call sites, I've updated `get_build_provider_flags` to not include `SNAPCRAFT_BUILD_INFO` if the option is turned off.  Without this, we'd set the environment variable to `"False"`, which be considered true by the various `if os.getenv("SNAPCRAFT_BUILD_INFO"):` calls.

cjp256 replied:
Good idea, works for the existing flags. Thanks!

jhenstridge commented:
I've updated the PR as suggested.  The CI seems to be failing for reasons unrelated to the changes in my branch.

jhenstridge commented:
Spread tests failed with:

```
2020-04-14 18:49:05 Error preparing google:ubuntu-18.04-64:tests/spread/general/build-aux (apr141843-278343) : 
-----
+ mkdir test-snap
+ cd test-snap
+ snapcraft init
Usage: snapcraft [OPTIONS] COMMAND [ARGS]...
Try 'snapcraft -h' for help.

Error: Invalid value for '--generate-manifest': on is not a valid boolean
```

I think I can fix this with a custom Click ParamType that does the simpler "anything non empty is true" interpretation we're using for the environment variable.

cjp256 commented:
@sergiusens should "on" be supported?  Interesting that it's used in spread.yaml, but I think everyone uses `SNAPCRAFT_BUILD_INFO=1`. I think we should fix it in our spread.yaml.

Unrelated, would `--enable-manifest` make better sense than `--generate-manifest`?

sergiusens commented:
I have no strong opinion on --enable versus --generate although your proposal has a better ring.

Wrt on, 1, true... we already make use of https://docs.python.org/3.6/distutils/apiref.html#distutils.util.strtobool in out code.

jhenstridge commented:
I've renamed the flag to `--enable-manifest`, and adjusted the parameter type to use `strtobool`.  I also updated the call sites making use of `SNAPCRAFT_BUILD_INFO` to use `strtobool` too, to keep things in sync.
